proposals/0570-string-interpolation.rst

```diff
@@ -0,0 +1,230 @@
+
+      textExample2 = T.pack $ T.unpack name1 <> " (age: " <> show age1 <> ") encountered " <> T.unpack name2 <> " (age: " <> show age2 <> ")"
+
+      textExample3 = T.unwords
+        [ name1
+        , "(age: " <> T.pack (show age1) <> ")"
+        , "encountered"
+        , name2
+        , "(age: " <> T.pack (show age2) <> ")"
+        ]
+
+* ``printf`` is partial and unsafe, which especially safety-conscious people might always stay away from anyway. Using a safer ``printf`` like ``formatting`` induces a third-party dependency, which is admittedly lightweight, but isn't as seamless of an integration as native string interpolation would be
+
+* Quasiquotes induces a dependency on Template Haskell, which a lot of people avoid out of principle. Most QuasiQuoters also add a dependency on ``haskell-src-exts`` to parse arbitrary Haskell expressions, which could technically be avoided by using something like ``ghc-meta`` (`repo <https://github.com/noughtmare/ghc-meta>`_, `GHC issue <https://gitlab.haskell.org/ghc/ghc/-/issues/20862>`_), but this isn't in wide use yet.
+
+If Haskell had native string interpolation, it would have the benefit and safety of the current third-party quasiquotes without the need for Template Haskell, and be able to take advantage of features like `multiline strings <https://github.com/ghc-proposals/ghc-proposals/pull/569>`_.
+::
+
+  error s"Expected: ${x + y}, got: ${result}"
+
+  let textExample = s"${name1} (age: ${age1}) encountered ${name2} (age: ${age2})"
+
+Proposed Change Specification
+-----------------------------
+
+When a new ``StringInterpolation`` extension is enabled, the following syntax:
+::
+
+  s"foo ${f a b} bar ${g x} baz ${name}"
+
+will be syntax sugar for:
+::
+
+  interpolateRaw "foo "#
+  . doInterpolate (f a b)
+  . interpolateRaw " bar "#
+  . doInterpolate (g x)
+  . interpolateRaw " baz "#
+  . doInterpolate name
+  $ interpolateEmpty
+
+where:
+::
+
+  class Interpolate s where
+    interpolateRaw :: Addr# -> s -> s
+    default interpolateRaw :: (IsString s, Semigroup s) => Addr# -> s -> s
+    interpolateRaw s = (fromString (unpackCString# s) <>)
+
+    interpolateEmpty :: s
+    default interpolateEmpty :: Monoid s => s
+    interpolateEmpty = mempty
+
+  class Interpolate s => InterpolateValueRunner s a where
+    doInterpolate :: a -> s -> s
+
+  class Interpolate s => InterpolateValue flag s a where
+    {-# MINIMAL interpolate | interpolatePrec #-}
+
+    interpolate :: Proxy flag -> a -> s
+    interpolate proxy a = interpolatePrec proxy 0 a interpolateEmpty
+
+    interpolatePrec :: Proxy flag -> Int -> a -> s -> s
+    default interpolatePrec :: Semigroup s => Proxy flag -> Int -> a -> s -> s
+    interpolatePrec proxy _ a s = interpolate proxy a <> s
+
+  data InterpolateDefault
+  data InterpolateOverride
+
+  class InterpolateValueFlag s a flag | s a -> flag
```

Comment:
I think what you're trying to do is to provide some customizable behavior for how values are interpolated in. I think we should at least discuss some simpler alternatives. The obvious ones to me are:
1. Values which are interpolated in must always be of the string type being interpolated. If you want to interpolate something else in you need to call `show` or whatever yourself. This is what Nix does, for example. This has the advantage of being very simple and giving the user full control just by calling functions. (This would be my preference)
2. Values which are interpolated in get `fromString . show` called on them. This is also very simple and relies on the language's default approach for displaying things, which is somewhat appropriate for a language feature.

It seems sad to add _another_ way of displaying strings that many users will need to implement. It's also unclear _who_ is going to end up writing these instances. Is it the author of the string type? Or is it the author of the type being interpolated? What if they disagree about how things should be put in?

Reply:
I guess if we did my suggestion for having a separate accumulator type then with alternative 1 we would probably want to allow either a string (in which case you call `fromString`) or a value of the accumulator type. Perhaps that's a little tricky.

Reply:
@michaelpj Lots of questions here :) I apologize if the proposal isn't clear, I can improve it more when I get a chance. Let me try to explain it from first principles.

I tried to imitate what Show does with showsPrec here for performance reasons, but let's ignore all of that. Fundamentally, I want interpolation to be variable in two ways:

1. The type of the final expression (analogous to IsString)
2. How a value should be converted into that type to be joined in (analogous to Show)

I want (1) to be different than just reusing IsString for things like an injection-safe SqlQuery type.

```hs
let val = "true; drop table tab;"
s"SELECT * FROM tab WHERE col = ${val}"

-- should result in:
SqlQuery
    "SELECT * FROM tab WHERE col = ?"
    [SqlString "true; drop table tab;"]
```

Ignoring performance, I want the following semantics:
```hs
s"a ${b} c"

-- assuming r is a Monoid
-- interpolateRaw :: String -> r
-- interpolateVal :: a -> r
interpolateRaw "a " <> interpolateVal b <> interpolateRaw " c"
```

I think you're suggesting Text.Builder as a way to solve the performance issue of the <>, but that's not the performance issue the Addr# was solving; the purpose of the Addr# is so interpolateRaw could go directly from `Addr# -> r`, for cases where that's a more optimal path than going `Addr# -> String -> r`.

I'm not sure how to respond to the rest of your comments. Does this explanation make sense? Do you mind rephrasing your comments in the context of this explanation?

Reply:
Oh and also: "Values get interpolated as `fromString . show`" won't work as-is, because you probably don't want strings to be quoted:

```hs
let foo = "abc"
putStrLn s"example: ${foo}"

-- example: "abc"
```

So you'd need at least a special case for String/Text to be `id`/`pack`/`unpack` instead of `show`

Reply:
> I want (1) to be different than just reusing IsString for things like an injection-safe SqlQuery type.

It would be helpful to explain all this in the motivation since it seems like a major design constraint! 

> the purpose of the Addr# is so interpolateRaw could go directly from Addr# -> r, for cases where that's a more optimal path than going Addr# -> String -> r

This seems unconvincing to me. Is the cost of processing the single string literals so big? I can see the case for worrying about repeated concatenation while building the interpolated string (since that can be accidentally quadratic), but processing the individual literal pieces doesn't seem like it should be any worse than processing string literals today, which I don't recall seeing performance complaints about?

Again, I think writing down the reasoning would help a lot!

> How a value should be converted into that type to be joined in (analogous to Show)

Okay, but you're baking in a whole new how-like mechanism to do that. If that's what you want, how about this?

```haskell
class (Semigroup (Accumulator s)) => Interpolate s where
   type Accumulator s :: Type
   type Hole s :: Type

   interpolateHole :: Hole s -> Accumulator s
   
   interpolateFinish :: Accumulator s -> s

instance Interpolate SqlQuery where
   type Accumulator SqlQuery = SqlQuery
   type Hole SqlQuery = SqlString
   
   interpolateHole = -- however you turn SqlStrings into queries
   interpolateFinish = id
```

The details are mostly just to be explicit, quite probably this doesn't quite work, but the idea is that we could also have a "hole" type, and then we can control how things get turned into holes (the user has to produce a value of the hole type), and also how the holes get folded into the accumulator.